src/v3/plugins/github/createGraph.js

```diff
@@ -49,113 +19,53 @@ class GraphCreator {
     this.graph = new Graph();
   }
 
-  addNode(addr: StructuredAddress) {
-    this.graph.addNode(toRaw(addr));
-  }
-
-  addData(data: GithubResponseJSON) {
-    this.addRepository(data.repository);
-  }
-
-  addRepository(repoJSON: RepositoryJSON) {
-    const repo: RepoAddress = {
-      type: "REPO",
-      owner: repoJSON.owner.login,
-      name: repoJSON.name,
-    };
-    this.addNode(repo);
-    repoJSON.issues.nodes.forEach((issue) => this.addIssue(repo, issue));
-    repoJSON.pulls.nodes.forEach((pull) => this.addPull(repo, pull));
-  }
-
-  addIssue(repo: RepoAddress, issueJSON: IssueJSON) {
-    const issue: IssueAddress = {
-      type: "ISSUE",
-      repo,
-      number: String(issueJSON.number),
-    };
-    this.addNode(issue);
-    this.addAuthors(issue, issueJSON.author);
-    this.addHasParent(issue, repo);
-    issueJSON.comments.nodes.forEach((comment) =>
-      this.addComment(issue, comment)
-    );
-  }
-
-  addPull(repo: RepoAddress, pullJSON: PullJSON) {
-    const pull: PullAddress = {
-      type: "PULL",
-      repo,
-      number: String(pullJSON.number),
-    };
-    this.addNode(pull);
-    this.addAuthors(pull, pullJSON.author);
-    this.addHasParent(pull, repo);
-    pullJSON.comments.nodes.forEach((c) => this.addComment(pull, c));
-    pullJSON.reviews.nodes.forEach((review) => this.addReview(pull, review));
-    if (pullJSON.mergeCommit != null) {
-      const commitHash = pullJSON.mergeCommit.oid;
-      const commit = {type: COMMIT_TYPE, hash: commitHash};
-      this.graph.addNode(gitToRaw(commit));
-      this.graph.addEdge(createEdge.mergedAs(pull, commit));
+  addData(view: R.RelationalView) {
+    function* entitiesInInsertionOrder() {
+      // Authors are added before authored content
+      // Parents are added before children
+      yield* view.userlikes();
+      yield* view.repos();
+      yield* view.issues();
+      yield* view.pulls();
+      yield* view.reviews();
+      yield* view.comments();
+    }
+    for (const entity of entitiesInInsertionOrder()) {
```

Comment:
Hmm. I’m not a huge fan of this structure. You mix all the data types
together into one heterogeneous stream, and then for each element case
on its type to perform some actions, possibly multiple depending on the
type. It seems pretty ad hoc and complicated compared to the following
simple approach:

```js
for (const user of view.userlikes()) {
  this.addNode(user.address());
}
for (const repo of view.repos()) {
  this.addNode(repo.address());
}
for (const issue of view.issues()) {
  this.addNode(issue.address());
  this.addHasParent(issue);
  this.addAuthors(issue);
}
for (const pull of view.pulls()) {
  this.addNode(pull.address());
  this.addHasParent(pull);
  this.addAuthors(pull);
  const commit = entity.mergedAs();
  if (commit != null) {
    this.graph.addNode(GitNode.toRaw(commit));
    this.graph.addEdge(createEdge.mergedAs(entity.address(), commit));
  }
}
for (const review of view.reviews()) {
  this.addNode(review.address());
  this.addHasParent(review);
  this.addAuthors(review);
}
for (const comment of view.comments()) {
  this.addNode(comment.address());
  this.addHasParent(comment);
  this.addAuthors(comment);
}
```

Compared to the approach above, the approach in this pull request has
less duplicated code, but it’s not clear that the deduplication actually
represents a useful abstraction.

For reference, the proposed approach is only two extra lines of code.

What do you think?

Reply:
I would alternately support a “trait-based” approach like:

```js
// Add nodes
for (const node of view.allEntities()) {
  this.addNode(node.address());
}

// Add parents
for (const node of view.childEntities()) {
  this.addHasParent(node);
}

// Add authors
for (const node of view.authorEntities()) {
  this.addAuthors(node);
}

// Add "merged-as" edges
for (const pull of view.pulls()) {
  const commit = entity.mergedAs();
  if (commit != null) {
    this.graph.addNode(GitNode.toRaw(commit));
    this.graph.addEdge(createEdge.mergedAs(entity.address(), commit));
  }
}
```

…with the differences from the approach in this pull request being that
(a) the entities come from named methods under the purview of the
relational view, (b) there’s clear separation among traits (e.g., the
coincidence of the set of child entities and the set of authorable
entities is not relevant), and (c) there’s no need to lose information
by injecting into a heterogeneous stream and then refine out of it
again.

Reply:
A notable difference between the two methods that I’ve proposed is how
they deal with extension. With the first method, it is easy to see how
to add a new entity type, whereas adding a new trait is a cross-cutting
concern. The opposite is true with the second proposal. In either case,
of course, adding a new entity type or trait is not _difficult_.

Reply:
Changed to your trait based approach.